### ISSUE TITLE
how can i pass through USB device in docker for windows(Hyper -v )

### ISSUE THREAD

**Issue by br13patel** [docker/for-win#3926]
Setup:

Docker for windows version:  18.09.2
Host: Windows 10, Version 1803 (OS build 17134.765)
Container: Linux/Ubuntu container.

Whenever i insert USB device it will not detect on my linux container. I have tried with many docker run command line option(e.g --device, -V : ), But still the behavior is the same.

USB device is detected if my host OS is Ubuntu and windows 7(with docker toolbox). I have also tried to debug this and dump kernel log but there is no log of my USB device.

**Comment by mikeparker** [comment]
Can you elaborate on exactly the reproduction steps? 
-> Plug in the USB device
-> Share the USB device from drive sharing inside Docker Desktop
-> A container cannot access the new drive?

**Comment by rajasec** [comment]
I have the same issue, not sure whether Hyper-V allows USB pass through due to which MobyVM not able to access the devices

**Comment by rajasec** [comment]
If it is Windows 10, I'm able to USB flash drive after applying the below steps

1) Click Docker Desktop settings
2) Click the new Drive and Apply.
3) After that launch the container by passing the drive name:
  For Eg: 
    docker run -ti -v D:/personal: /data ubuntu bash
With this I'm able to access the flash drive

**Comment by 1138-4EB** [comment]
For USB devices which are not for storage, USB over network might be a usable workaround until sharing the devices is supported by default. See https://github.com/ghdl/docker/tree/feat-usbip/usbip.

Ref #4105.

**Comment by androniychuk** [comment]
According to this doc, it's not supported
https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/hardware-devices-in-containers

But technically since you pass an id into a windows container, you should be able to do the same with Linux. Just need to figure out how the backend works.

**Comment by amirhs712** [comment]
I could really use this feature if it was added.
I have a .NET project which requires access to a security token via USB, but It won't work on windows containers.

**Comment by mschweig** [comment]
I would also need this feature. 
Windows 10 Host and Ubuntu Container

**Comment by petriceko** [comment]
No update?

**Comment by mfiumara** [comment]
Hey I think this is a limitation of windows not necessarily docker so until windows fixes USB devices for WSL you will have to do with usb over IP solutions (which are far from ideal).

**Comment by JPHutchins** [comment]
I am researching the use case "firmware development" and since USB passthrough is necessary for attaching programmers and serial I am interpreting that my best approach with a Windows Host will be a hypervisor with good USB passthrough to a Linux VM that then hosts the Docker containers.  

USB Windows->VMWare->LinuxVM->Docker->DockerContainer
SSH Windows->DockerContainer

Is this a reasonable interpretation of the state until Windows Docker Desktop benefits from a hypothetical WSL update?

**Comment by eine** [comment]
@JPHutchins it is. However, I'm not sure you can have an hypervisor with good USB passthrough. It that was feasible, I would have expected either Docker or Microsoft to provide that capability already.

Alternatively, you might want to patch the LinuxVM for supporting USB/IP. See https://hdl.github.io/containers/ug/index.html#_usbip_protocol_support_for_docker_desktop. The challenge in this case is finding an open source USB/IP tool on Windows, but it works ok if you are good with using a non open source tool.

**Comment by JPHutchins** [comment]
Thanks!  I am already happy with USB passthrough on Windows hosts using VMWare as well us on Ubuntu hosts using KVM so I hope that adding Docker at the Ubuntu VM layer is just as happy.

**Comment by davidnaumann-bastian** [comment]
Would recommend this approach: [Microsoft Dev Blog: Connecting USB devices to WSL](https://devblogs.microsoft.com/commandline/connecting-usb-devices-to-wsl/ ) and just pick up those USB devices in WSL2.

I'd say until Microsoft official support pushing USB devices down to the WSL level this is the best and only approach as the time of writing this.

**Comment by znmeb** [comment]
> Would recommend this approach: [Microsoft Dev Blog: Connecting USB devices to WSL](https://devblogs.microsoft.com/commandline/connecting-usb-devices-to-wsl/) and just pick up those USB devices in WSL2.
> 
> I'd say until Microsoft officially supports pushing USB devices down to the WSL level this is the best and only approach as the time of writing this.

USB support in WSL is poorly documented and not solid. It does not as of this writing solve the problem.

**Comment by davidnaumann-bastian** [comment]
> > Would recommend this approach: [Microsoft Dev Blog: Connecting USB devices to WSL](https://devblogs.microsoft.com/commandline/connecting-usb-devices-to-wsl/) and just pick up those USB devices in WSL2.
> > I'd say until Microsoft officially supports pushing USB devices down to the WSL level this is the best and only approach as the time of writing this.
> 
> USB support in WSL is poorly documented and not solid. It does not as of this writing solve the problem.

Yes, agreed with the sentiment expressed. Should've added that the only reason I recommended that approach is that it is the only approach I found to go about this problem.

**Comment by AShriki** [comment]
I see the original poster has not elaborated on your request for information and has probably abandoned this issue. My specific use case is to generalize embedded development environments with a debugger (see segger j-link for an example [here](https://www.segger.com/products/debug-probes/j-link/)). These devices use vendor supplied drivers. I would also be interested in passing through any generic USB device similarly to how you can pass a USB device from the host into a VirtualBox guest OS ([this guide](https://www.eltima.com/article/virtualbox-usb-passthrough/) is a simplistic overview in general and not specific to a windows host).

`> Can you elaborate on exactly the reproduction steps? `
`-> Plug in the USB device `
The J-link usb device is plugged into a USB port on the computer
`-> A container cannot access the new drive?`
Not relevant, but no USB device is seen in /dev/ and I can't find a way to allow the docker container to use/take ownership of the device.

What I would like is a way to specify a USB device connected to the host for use (exclusive or not. No preference here) within the container.

Let me know what other information is needed.

**Comment by androniychuk** [comment]
Would like to post my thoughts on this.
I originally needed such an implementation for the purpose of Ubuntu running on WSL and accessing the graphics driver. However, there is now another implementation of this, that works great. Its the WSL driver for your GPU:
https://learn.microsoft.com/en-us/windows/wsl/tutorials/gui-apps
This doesn't work on all systems, but it works pretty great on my Intel GPU and accomplishes the "pass-through" that I needed.

**Comment by beriberikix** [comment]
We now have a potential solution in the form of USB/IP. I wrote up details [here](https://blog.golioth.io/usb-docker-windows-macos/).
